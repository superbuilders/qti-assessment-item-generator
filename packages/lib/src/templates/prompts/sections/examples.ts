import { existsSync, readFileSync } from "node:fs"
import * as path from "node:path"
import { resolveLibPath } from "@/internal/paths"

const POSITIVE_EXAMPLE_CACHE = new Map<string, string>()
const NEGATIVE_EXAMPLE_CACHE = new Map<string, string>()

const POSITIVE_ROOT = resolveLibPath("templates/prompts/examples/positive")
const POSITIVE_TEMPLATES_DIR = path.join(POSITIVE_ROOT, "templates")
const POSITIVE_NOTES_DIR = path.join(POSITIVE_ROOT, "notes")

const NEGATIVE_ROOT = resolveLibPath("templates/prompts/examples/negative")
const NEGATIVE_TEMPLATES_DIR = path.join(NEGATIVE_ROOT, "templates")
const NEGATIVE_NOTES_DIR = path.join(NEGATIVE_ROOT, "notes")

export function createPositiveExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadPositiveExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### POSITIVE_EXAMPLES
<examples type="positive">
${rendered}
</examples>`
}

export function createNegativeExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadNegativeExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### NEGATIVE_EXAMPLES
<examples type="negative">
${rendered}
</examples>`
}

function loadPositiveExample(name: string): string {
	const cached = POSITIVE_EXAMPLE_CACHE.get(name)
	if (cached) return cached

	const codePath = path.join(POSITIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}

	const notesPath = path.join(POSITIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""

	const rendered = `<example kind="positive" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	POSITIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}

function loadNegativeExample(name: string): string {
	const existing = NEGATIVE_EXAMPLE_CACHE.get(name)
	if (existing) return existing

	const codePath = path.join(NEGATIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}
	const notesPath = path.join(NEGATIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""
	const rendered = `<example kind="negative" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	NEGATIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}
