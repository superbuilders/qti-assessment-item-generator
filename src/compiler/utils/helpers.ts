import { typedSchemas } from "@/widgets/registry"

/**
 * Calculates the greatest common divisor (GCD) of two numbers using the Euclidean algorithm.
 * @param a The first number.
 * @param b The second number.
 * @returns The GCD of a and b.
 */
export function gcd(a: number, b: number): number {
	return b === 0 ? a : gcd(b, a % b)
}

/**
 * Determines if a fraction results in a terminating decimal.
 * A fraction terminates if and only if the prime factors of its simplified denominator are only 2s and 5s.
 * @param numerator The numerator of the fraction.
 * @param denominator The denominator of the fraction.
 * @returns True if the fraction terminates, false otherwise.
 */
export function isTerminatingFraction(
	numerator: number,
	denominator: number
): boolean {
	if (denominator === 0) {
		return false
	}
	// Simplify the fraction by dividing by the GCD.
	const commonDivisor = gcd(Math.abs(numerator), Math.abs(denominator))
	let simplifiedDen = Math.abs(denominator) / commonDivisor

	// The fraction terminates if the simplified denominator's prime factors are only 2 and 5.
	while (simplifiedDen % 2 === 0) {
		simplifiedDen /= 2
	}
	while (simplifiedDen % 5 === 0) {
		simplifiedDen /= 5
	}

	return simplifiedDen === 1
}

export function encodeDataUri(content: string): string {
	const base64 = btoa(unescape(encodeURIComponent(content)))
	const isSvg = content.trim().startsWith("<svg")
	return `${isSvg ? "data:image/svg+xml" : "data:text/html"};base64,${base64}`
}

// Helper function to check if a string is a valid widget type
export function isValidWidgetType(
	type: string
): type is keyof typeof typedSchemas {
	return type in typedSchemas
}

/**
 * Attempts to approximate simple multiplicative/divisive expressions that include π or pi
 * by replacing pi with the provided approximation (default 3.14) and evaluating the expression.
 *
 * Supported forms (case-insensitive for "pi"):
 * - "144*pi", "pi*144", "144pi" (implicit multiplication), "pi144" (rare, but treated as pi*144)
 * - Chains with '*' and '/' only, e.g. "3/2*pi", "pi/4", "6*pi/5", "(144) * (pi)"
 * Parentheses are allowed but only for grouping; no addition/subtraction is supported.
 *
 * Returns a normalized decimal string if evaluation succeeds, otherwise null.
 */
export function tryApproximatePiProduct(
	rawExpr: string,
	piApprox = 3.14
): string | null {
	if (typeof rawExpr !== "string" || rawExpr.trim() === "") return null

	// Quick check: must contain a pi token of some sort
	const containsPi = /\bpi\b/i.test(rawExpr) || rawExpr.includes("π")
	if (!containsPi) return null

	// Normalize: insert explicit * for implicit numeric-pi adjacency in common cases
	const expr = rawExpr
		// insert * between number and pi (e.g., 144pi -> 144*pi)
		.replace(/(\d)\s*(π|\bpi\b)/gi, "$1*$2")
		// and between pi and number (e.g., pi144 -> pi*144)
		.replace(/(π|\bpi\b)\s*(\d)/gi, "$1*$2")
		// remove whitespace
		.replace(/\s+/g, "")
		// drop parentheses (multiplicative-only evaluation does not require them)
		.replace(/[()]/g, "")
		// normalize 'pi' variants to numeric approximation
		.replace(/π/gi, String(piApprox))
		.replace(/\bpi\b/gi, String(piApprox))

	// After substitution, we only allow digits, '.', '*', '/' characters
	if (/[^0-9.*/]/.test(expr)) return null
	// Disallow addition/subtraction in this lightweight evaluator
	if (/[+-]/.test(expr)) return null
	// Must still contain the approximation value to ensure we actually used pi
	if (!expr.includes(String(piApprox))) return null

	// Split into tokens by * and / and evaluate left-to-right (* and / are same precedence)
	const parts: string[] = expr.split(/([*/])/)
	if (parts.length === 0) return null

	// Validate tokens
	for (let i = 0; i < parts.length; i++) {
		const t = parts[i]
		if (i % 2 === 0) {
			// operand
			if (t === "" || Number.isNaN(Number(t))) return null
		} else {
			// operator must be * or /
			if (t !== "*" && t !== "/") return null
		}
	}

	let acc = Number(parts[0])
	for (let i = 1; i < parts.length; i += 2) {
		const op = parts[i]
		const rhs = Number(parts[i + 1])
		if (op === "*") acc = acc * rhs
		else {
			if (rhs === 0) return null
			acc = acc / rhs
		}
	}

	// Normalize number to a compact string while preserving significant digits
	// Avoid scientific notation for typical classroom values
	const normalizeNumberString = () => {
		const maybeFixed = acc.toFixed(10) // limit noise
		// trim trailing zeros
		const trimmed = maybeFixed.replace(/\.0+$/, "").replace(/(\.\d*?)0+$/, "$1")
		return trimmed
	}
	const asString = normalizeNumberString()

	// Add a guard: must be a finite number
	if (!Number.isFinite(Number(asString))) return null
	return asString
}

/**
 * Normalizes a string part for use in a derived QTI identifier.
 * Converts to upper snake case and removes invalid characters.
 * @param part The string part to normalize.
 * @returns The normalized string.
 */
export function normalizeIdPart(part: string): string {
	return part.toUpperCase().replace(/[^A-Z0-9_]/g, "_")
}

/**
 * Derives a deterministic, QTI-safe combo identifier from a path of response parts.
 * @param pathParts An array of strings, where each string is a response-value pair (e.g., "RESPONSE_1_A").
 * @returns The fully-formed combo identifier (e.g., "FB__RESPONSE_1_A__RESPONSE_2_CORRECT").
 */
export function deriveComboIdentifier(pathParts: string[]): string {
	return `FB__${pathParts.join("__")}`
}
