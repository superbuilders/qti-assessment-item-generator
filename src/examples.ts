import type { AssessmentItemInput } from "@/core/item"
import type { allWidgetsCollection } from "@/widgets/collections/all"
import type { WidgetTypeTupleFrom } from "@/widgets/collections/types"

export const allExamples: AssessmentItemInput<WidgetTypeTupleFrom<typeof allWidgetsCollection>>[] = []
