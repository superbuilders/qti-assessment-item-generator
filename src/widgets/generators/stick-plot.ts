import * as errors from "@superbuilders/errors"
import * as logger from "@superbuilders/slog"
import { z } from "zod"
import type { WidgetGenerator } from "../types"
import { CanvasImpl } from "../utils/canvas-impl"
import { AXIS_VIEWBOX_PADDING } from "../utils/constants"
import { setupCoordinatePlaneBaseV2 } from "../utils/coordinate-plane-utils"
import { CSS_COLOR_PATTERN } from "../utils/css-color"
import { abbreviateMonth } from "../utils/labels"
import { createHeightSchema, createWidthSchema } from "../utils/schemas"
import { theme } from "../utils/theme"

// A single vertical stick at an x-category with a height determined by y value
const StickSchema = z
	.object({
		xLabel: z.string().describe("Label for the position along the x-axis (e.g., isotope mass label)."),
		yValue: z.number().nonnegative().describe("Height of the stick along the y-axis. Typically a percentage or count."),
		color: z
			.string()
			.regex(CSS_COLOR_PATTERN, "invalid css color; use hex (#RGB, #RRGGBB, #RRGGBBAA)")
			.describe("Hex-only color for the stick stroke.")
	})
	.strict()

// Optional reference lines such as mean/average
const ReferenceLineSchema = z
	.object({
		xLabel: z.string().describe("Category label at which to draw a vertical reference line."),
		label: z.string().describe("Text to annotate near the reference line (e.g., 'Average')."),
		color: z
			.string()
			.regex(CSS_COLOR_PATTERN, "invalid css color; use hex (#RGB, #RRGGBB, #RRGGBBAA)")
			.describe("Color of the reference line and label.")
	})
	.strict()

export const StickPlotPropsSchema = z
	.object({
		type: z.literal("stickPlot"),
		width: createWidthSchema(),
		height: createHeightSchema(),
		title: z.string().describe("Main title displayed above the plot."),
		xAxis: z
			.object({
				label: z.string().describe("Label for the horizontal axis (e.g., 'Atomic mass (u)')."),
				categories: z
					.array(z.string().min(1))
					.min(1)
					.describe("Complete ordered list of category labels along the x-axis."),
				showGridLines: z.boolean().describe("Whether to display vertical grid lines.")
			})
			.strict(),
		yAxis: z
			.object({
				label: z.string().describe("Label for the vertical axis (e.g., 'Relative abundance (%)')."),
				min: z.number().describe("Minimum numeric value on the y-axis."),
				max: z.number().describe("Maximum numeric value on the y-axis."),
				tickInterval: z.number().positive().describe("Spacing between y-axis ticks."),
				showGridLines: z.boolean().describe("Whether to display horizontal grid lines.")
			})
			.strict(),
		sticks: z.array(StickSchema),
		stickWidthPx: z.number().positive(),
		references: z.array(ReferenceLineSchema)
	})
	.strict()
	.describe(
		"Renders a categorical stick plot (mass spectrum-like). Each category can have a vertical 'stick' whose height represents an abundance or count."
	)

export type StickPlotProps = z.infer<typeof StickPlotPropsSchema>

export const generateStickPlot: WidgetGenerator<typeof StickPlotPropsSchema> = async (props) => {
	const { width, height, title, xAxis, yAxis, sticks, stickWidthPx, references } = props

	// Validate that all sticks map to known categories and values are within range
	const categoryToIndex = new Map<string, number>()
	for (const [i, c] of xAxis.categories.entries()) {
		categoryToIndex.set(c, i)
	}
	for (const s of sticks) {
		if (!categoryToIndex.has(s.xLabel)) {
			logger.error("stick xLabel missing from categories", { xLabel: s.xLabel })
			throw errors.new("stick plot: xLabel must exist in xAxis.categories")
		}
		if (s.yValue < yAxis.min || s.yValue > yAxis.max) {
			logger.error("stick yValue out of bounds", { y: s.yValue, min: yAxis.min, max: yAxis.max })
			throw errors.new("stick plot: yValue out of y-axis bounds")
		}
	}

	const canvas = new CanvasImpl({
		chartArea: { left: 0, top: 0, width, height },
		fontPxDefault: 12,
		lineHeightDefault: 1.2
	})

	// Use categoryBand for x, numeric for y
	const baseInfo = setupCoordinatePlaneBaseV2(
		{
			width,
			height,
			title,
			xAxis: {
				xScaleType: "categoryBand",
				label: xAxis.label,
				categories: xAxis.categories.map((c) => abbreviateMonth(c)),
				showTickLabels: true,
				showGridLines: xAxis.showGridLines
			},
			yAxis: {
				label: yAxis.label,
				min: yAxis.min,
				max: yAxis.max,
				tickInterval: yAxis.tickInterval,
				showTickLabels: true,
				showGridLines: yAxis.showGridLines
			}
		},
		canvas
	)

	// Render sticks within clipped chart area
	const bandWidth = baseInfo.bandWidth
	if (bandWidth === undefined) {
		logger.error("stick plot requires bandWidth for categorical x-axis", { categoriesCount: xAxis.categories.length })
		throw errors.new("stick plot: missing bandWidth for categorical x-axis")
	}

	const halfStick = Math.min(stickWidthPx, Math.max(1, bandWidth * 0.9)) / 2

	canvas.drawInClippedRegion((clippedCanvas) => {
		for (const s of sticks) {
			const i = categoryToIndex.get(s.xLabel)
			if (i === undefined) continue
			const cx = baseInfo.toSvgX(i)
			const yBottom = baseInfo.toSvgY(Math.max(yAxis.min, 0))
			const yTop = baseInfo.toSvgY(s.yValue)

			// Draw vertical line as the stick
			clippedCanvas.drawLine(cx, yBottom, cx, yTop, {
				stroke: s.color,
				strokeWidth: stickWidthPx
			})

			// Cap marker at the top for clarity
			clippedCanvas.drawLine(cx - halfStick, yTop, cx + halfStick, yTop, {
				stroke: s.color,
				strokeWidth: Math.max(1, stickWidthPx - 1)
			})
		}
	})

	// Optional reference lines at categories
	for (const ref of references) {
		const idx = categoryToIndex.get(ref.xLabel)
		if (idx === undefined) {
			logger.error("reference xLabel missing from categories", { xLabel: ref.xLabel })
			throw errors.new("stick plot: reference xLabel must exist in xAxis.categories")
		}
		const x = baseInfo.toSvgX(idx)
		canvas.drawLine(x, baseInfo.chartArea.top, x, baseInfo.chartArea.top + baseInfo.chartArea.height, {
			stroke: ref.color,
			strokeWidth: theme.stroke.width.thin
		})
		canvas.drawText({
			x: x + 4,
			y: baseInfo.chartArea.top + 12,
			text: abbreviateMonth(ref.label),
			fill: ref.color,
			anchor: "start"
		})
	}

	const { svgBody, vbMinX, vbMinY, width: finalWidth, height: finalHeight } = canvas.finalize(AXIS_VIEWBOX_PADDING)
	return `<svg width="${finalWidth}" height="${finalHeight}" viewBox="${vbMinX} ${vbMinY} ${finalWidth} ${finalHeight}" xmlns="http://www.w3.org/2000/svg" font-family="${theme.font.family.sans}" font-size="${theme.font.size.base}">${svgBody}</svg>`
}
