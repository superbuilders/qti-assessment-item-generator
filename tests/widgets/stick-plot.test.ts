import { describe, expect, test } from "bun:test"
import * as errors from "@superbuilders/errors"
import * as logger from "@superbuilders/slog"
import { stickPlotExamples } from "../../examples/stick-plot"
import type { WidgetInput } from "../../src/widgets/registry"
import { generateWidgetForTest } from "../helpers/generateWidgetForTest"

describe("Widget: stick-plot", () => {
	const examples: WidgetInput[] = stickPlotExamples

	examples.forEach((props, index) => {
		test(`should produce consistent output for example #${index + 1}`, async () => {
			const result = await errors.try(generateWidgetForTest(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error, index })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})
	})
})
