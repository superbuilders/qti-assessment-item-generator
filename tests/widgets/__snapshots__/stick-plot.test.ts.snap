// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Widget: stick-plot should produce consistent output for example #1 1`] = `"<svg width="619" height="541" viewBox="-99 -57 619 541" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><clipPath id="clip-0"><rect x="0" y="0" width="500" height="400"/></clipPath></defs><text x="250" y="-36.599999999999994" font-size="18" text-anchor="middle" fill="#333333" dominant-baseline="hanging"><tspan x="250" dy="0em">Mass Spectrum of Chlorine</tspan></text><line x1="0" y1="0" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="-5" y1="400" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><text x="-8" y="404" font-size="12" text-anchor="end" fill="#333333">0</text><line x1="0" y1="400" x2="500" y2="400" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="320" x2="0" y2="320" stroke="#333333" stroke-width="1.5"/><text x="-8" y="324" font-size="12" text-anchor="end" fill="#333333">20</text><line x1="0" y1="320" x2="500" y2="320" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="240" x2="0" y2="240" stroke="#333333" stroke-width="1.5"/><text x="-8" y="244" font-size="12" text-anchor="end" fill="#333333">40</text><line x1="0" y1="240" x2="500" y2="240" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="160" x2="0" y2="160" stroke="#333333" stroke-width="1.5"/><text x="-8" y="164" font-size="12" text-anchor="end" fill="#333333">60</text><line x1="0" y1="160" x2="500" y2="160" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="80" x2="0" y2="80" stroke="#333333" stroke-width="1.5"/><text x="-8" y="84" font-size="12" text-anchor="end" fill="#333333">80</text><line x1="0" y1="80" x2="500" y2="80" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="0" x2="0" y2="0" stroke="#333333" stroke-width="1.5"/><text x="-8" y="4" font-size="12" text-anchor="end" fill="#333333">100</text><line x1="0" y1="0" x2="500" y2="0" stroke="#e0e0e0" stroke-width="1"/><text x="-68.60000000000001" y="200" font-size="16" text-anchor="middle" fill="#333333" dominant-baseline="middle" transform="rotate(-90, -68.60000000000001, 200)"><tspan x="-68.60000000000001" dy="0em">Relative abundance (%)</tspan></text><line x1="0" y1="400" x2="500" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="50" y1="400" x2="50" y2="405" stroke="#333333" stroke-width="1.5"/><text x="50" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">35</text><line x1="150" y1="400" x2="150" y2="405" stroke="#333333" stroke-width="1.5"/><text x="150" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">36</text><line x1="250" y1="400" x2="250" y2="405" stroke="#333333" stroke-width="1.5"/><text x="250" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">37</text><line x1="350" y1="400" x2="350" y2="405" stroke="#333333" stroke-width="1.5"/><text x="350" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">38</text><line x1="450" y1="400" x2="450" y2="405" stroke="#333333" stroke-width="1.5"/><text x="450" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">39</text><text x="250" y="447" font-size="16" text-anchor="middle" fill="#333333"><tspan x="250" dy="0em">Atomic mass (u)</tspan></text><g clip-path="url(#clip-0)"><line x1="50" y1="400" x2="50" y2="96.80000000000001" stroke="#2563eb" stroke-width="4"/><line x1="48" y1="96.80000000000001" x2="52" y2="96.80000000000001" stroke="#2563eb" stroke-width="3"/><line x1="250" y1="400" x2="250" y2="303.2" stroke="#2563eb" stroke-width="4"/><line x1="248" y1="303.2" x2="252" y2="303.2" stroke="#2563eb" stroke-width="3"/></g></svg>"`;

exports[`Widget: stick-plot should produce consistent output for example #2 1`] = `"<svg width="712" height="491" viewBox="-92 -57 712 491" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><clipPath id="clip-0"><rect x="0" y="0" width="600" height="350"/></clipPath></defs><text x="300" y="-36.599999999999994" font-size="18" text-anchor="middle" fill="#333333" dominant-baseline="hanging"><tspan x="300" dy="0em">Isotopic Distribution of Bromine</tspan></text><line x1="0" y1="0" x2="0" y2="350" stroke="#333333" stroke-width="1.5"/><line x1="-5" y1="350" x2="0" y2="350" stroke="#333333" stroke-width="1.5"/><text x="-8" y="354" font-size="12" text-anchor="end" fill="#333333">0</text><line x1="0" y1="350" x2="600" y2="350" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="291.6666666666667" x2="0" y2="291.6666666666667" stroke="#333333" stroke-width="1.5"/><text x="-8" y="295.6666666666667" font-size="12" text-anchor="end" fill="#333333">10</text><line x1="0" y1="291.6666666666667" x2="600" y2="291.6666666666667" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="233.33333333333334" x2="0" y2="233.33333333333334" stroke="#333333" stroke-width="1.5"/><text x="-8" y="237.33333333333334" font-size="12" text-anchor="end" fill="#333333">20</text><line x1="0" y1="233.33333333333334" x2="600" y2="233.33333333333334" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="175" x2="0" y2="175" stroke="#333333" stroke-width="1.5"/><text x="-8" y="179" font-size="12" text-anchor="end" fill="#333333">30</text><line x1="0" y1="175" x2="600" y2="175" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="116.66666666666669" x2="0" y2="116.66666666666669" stroke="#333333" stroke-width="1.5"/><text x="-8" y="120.66666666666669" font-size="12" text-anchor="end" fill="#333333">40</text><line x1="0" y1="116.66666666666669" x2="600" y2="116.66666666666669" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="58.333333333333314" x2="0" y2="58.333333333333314" stroke="#333333" stroke-width="1.5"/><text x="-8" y="62.333333333333314" font-size="12" text-anchor="end" fill="#333333">50</text><line x1="0" y1="58.333333333333314" x2="600" y2="58.333333333333314" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="0" x2="0" y2="0" stroke="#333333" stroke-width="1.5"/><text x="-8" y="4" font-size="12" text-anchor="end" fill="#333333">60</text><line x1="0" y1="0" x2="600" y2="0" stroke="#e0e0e0" stroke-width="1"/><text x="-61.6" y="175" font-size="16" text-anchor="middle" fill="#333333" dominant-baseline="middle" transform="rotate(-90, -61.6, 175)"><tspan x="-61.6" dy="0em">Intensity</tspan></text><line x1="0" y1="350" x2="600" y2="350" stroke="#333333" stroke-width="1.5"/><line x1="75" y1="350" x2="75" y2="355" stroke="#333333" stroke-width="1.5"/><text x="75" y="363" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">79</text><line x1="225" y1="350" x2="225" y2="355" stroke="#333333" stroke-width="1.5"/><text x="225" y="363" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">80</text><line x1="375" y1="350" x2="375" y2="355" stroke="#333333" stroke-width="1.5"/><text x="375" y="363" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">81</text><line x1="525" y1="350" x2="525" y2="355" stroke="#333333" stroke-width="1.5"/><text x="525" y="363" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">82</text><text x="300" y="397" font-size="16" text-anchor="middle" fill="#333333"><tspan x="300" dy="0em">Mass (amu)</tspan></text><g clip-path="url(#clip-0)"><line x1="75" y1="350" x2="75" y2="54.24999999999994" stroke="#dc2626" stroke-width="6"/><line x1="72" y1="54.24999999999994" x2="78" y2="54.24999999999994" stroke="#dc2626" stroke-width="5"/><line x1="375" y1="350" x2="375" y2="62.416666666666686" stroke="#dc2626" stroke-width="6"/><line x1="372" y1="62.416666666666686" x2="378" y2="62.416666666666686" stroke="#dc2626" stroke-width="5"/></g><line x1="225" y1="0" x2="225" y2="350" stroke="#059669" stroke-width="1"/><text x="229" y="12" font-size="12" fill="#059669">Average</text></svg>"`;

exports[`Widget: stick-plot should produce consistent output for example #3 1`] = `"<svg width="562" height="441" viewBox="-92 -57 562 441" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><clipPath id="clip-0"><rect x="0" y="0" width="450" height="300"/></clipPath></defs><text x="225" y="-36.599999999999994" font-size="18" text-anchor="middle" fill="#333333" dominant-baseline="hanging"><tspan x="225" dy="0em">Elemental Analysis Results</tspan></text><line x1="0" y1="0" x2="0" y2="300" stroke="#333333" stroke-width="1.5"/><line x1="-5" y1="300" x2="0" y2="300" stroke="#333333" stroke-width="1.5"/><text x="-8" y="304" font-size="12" text-anchor="end" fill="#333333">0</text><line x1="0" y1="300" x2="450" y2="300" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="240" x2="0" y2="240" stroke="#333333" stroke-width="1.5"/><text x="-8" y="244" font-size="12" text-anchor="end" fill="#333333">10</text><line x1="0" y1="240" x2="450" y2="240" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="180" x2="0" y2="180" stroke="#333333" stroke-width="1.5"/><text x="-8" y="184" font-size="12" text-anchor="end" fill="#333333">20</text><line x1="0" y1="180" x2="450" y2="180" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="120" x2="0" y2="120" stroke="#333333" stroke-width="1.5"/><text x="-8" y="124" font-size="12" text-anchor="end" fill="#333333">30</text><line x1="0" y1="120" x2="450" y2="120" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="60" x2="0" y2="60" stroke="#333333" stroke-width="1.5"/><text x="-8" y="64" font-size="12" text-anchor="end" fill="#333333">40</text><line x1="0" y1="60" x2="450" y2="60" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="0" x2="0" y2="0" stroke="#333333" stroke-width="1.5"/><text x="-8" y="4" font-size="12" text-anchor="end" fill="#333333">50</text><line x1="0" y1="0" x2="450" y2="0" stroke="#e0e0e0" stroke-width="1"/><text x="-61.6" y="150" font-size="16" text-anchor="middle" fill="#333333" dominant-baseline="middle" transform="rotate(-90, -61.6, 150)"><tspan x="-61.6" dy="0em">Percentage (%)</tspan></text><line x1="0" y1="300" x2="450" y2="300" stroke="#333333" stroke-width="1.5"/><line x1="45" y1="300" x2="45" y2="305" stroke="#333333" stroke-width="1.5"/><text x="45" y="313" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">H</text><line x1="135" y1="300" x2="135" y2="305" stroke="#333333" stroke-width="1.5"/><text x="135" y="313" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">C</text><line x1="225" y1="300" x2="225" y2="305" stroke="#333333" stroke-width="1.5"/><text x="225" y="313" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">N</text><line x1="315" y1="300" x2="315" y2="305" stroke="#333333" stroke-width="1.5"/><text x="315" y="313" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">O</text><line x1="405" y1="300" x2="405" y2="305" stroke="#333333" stroke-width="1.5"/><text x="405" y="313" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">S</text><text x="225" y="347" font-size="16" text-anchor="middle" fill="#333333"><tspan x="225" dy="0em">Element</tspan></text><g clip-path="url(#clip-0)"><line x1="45" y1="300" x2="45" y2="250.8" stroke="#7c3aed" stroke-width="8"/><line x1="41" y1="250.8" x2="49" y2="250.8" stroke="#7c3aed" stroke-width="7"/><line x1="135" y1="300" x2="135" y2="47.39999999999998" stroke="#7c3aed" stroke-width="8"/><line x1="131" y1="47.39999999999998" x2="139" y2="47.39999999999998" stroke="#7c3aed" stroke-width="7"/><line x1="225" y1="300" x2="225" y2="206.4" stroke="#7c3aed" stroke-width="8"/><line x1="221" y1="206.4" x2="229" y2="206.4" stroke="#7c3aed" stroke-width="7"/><line x1="315" y1="300" x2="315" y2="103.20000000000002" stroke="#7c3aed" stroke-width="8"/><line x1="311" y1="103.20000000000002" x2="319" y2="103.20000000000002" stroke="#7c3aed" stroke-width="7"/><line x1="405" y1="300" x2="405" y2="292.2" stroke="#7c3aed" stroke-width="8"/><line x1="401" y1="292.2" x2="409" y2="292.2" stroke="#7c3aed" stroke-width="7"/></g></svg>"`;
